### PR TITLE
fix cookie extraction when http/2 is used

### PR DESCRIPTION
**Pull request by valkum**
With http/2 the spec allows a single Cookie header for each cookie for better compression. The current code only checks the first cookie header and subsequent headers are not checked. This changes the behaviour. See tests.

### CODE REVIEW

File: mendes/src/cookies.rs [hunk 1]
```diff
@@ -145,25 +145,27 @@ where
...
     let name = T::NAME;
-    for cookie in cookies.split(';') {
-        let cookie = cookie.trim_start();
-        if cookie.len() < (name.len() + 1 + NONCE_LEN + TAG_LEN)
-            || !cookie.starts_with(name)
-            || cookie.as_bytes()[name.len()] != b'='
-        {
-            continue;
-        }
+    for cookie_header in cookies {
```
Comment (djc) [inline mendes/src/cookies.rs:150]: I think I'd prefer to call `cookie_header` `value` instead, and we can probably just inline the call to `headers.get_all()` in the loop header?

If we fail to decode the UTF-8, we should continue rather than returning `None`:

```rust
let Ok(value) = str::from_utf8(value.as_ref()) {
    continue;
};
```

File: mendes/src/cookies.rs [hunk 2]
```diff
@@ -231,3 +233,74 @@ pub enum Error {
...
+    #[derive(Clone, Copy, Debug, Deserialize, Serialize)]
+    pub struct Session {
+        id: i64,
+    }
+
+    impl super::CookieData for Session {
+        const NAME: &'static str = "Session";
+    }
+
+    fn value<T: CookieData>(key: &Key, data: T) -> Result<String, Error> {
```
Comment (djc) [inline mendes/src/cookies.rs:296]: This seems like a good reason to split the existing `store()` function in a different way to match what is needed here. Could hoist the `cookie::<T>(Some(&value))` out into `set_cookie_header()` and maybe respin the rest of `store()` into `Cookie::encode(data: T, key: &Key) -> Result<String, Error>`?

**Review by djc — COMMENTED** [review]
Good work!

File: mendes/src/cookies.rs [hunk 3]
```diff
@@ -144,40 +144,44 @@ where
     data: T,
 }
 
+impl<T: CookieData> Cookie<T> {
```
Comment (djc) [inline mendes/src/cookies.rs:147]: Would be nice to move this into a separate commit.

File: mendes/src/cookies.rs [hunk 4]
```diff
@@ -231,3 +235,63 @@ pub enum Error {
...
 }
+
+#[cfg(test)]
+mod test {
+    use http::{header, HeaderMap};
+    use serde::{Deserialize, Serialize};
+
+    use super::*;
+
+    /// This test checks that we can extract a cookie from a request that follows the HTTP/1.1 spec
```
Comment (djc) [inline mendes/src/cookies.rs:246]: Let's change the comments here to refer to just "separate headers" vs "multiple cookies in a single header" and avoid referring to the versions?

File: mendes/src/cookies.rs [hunk 5]
```diff
@@ -144,40 +144,44 @@ where
...
-    let cookies = str::from_utf8(cookies.as_ref()).ok()?;
     let name = T::NAME;
-    for cookie in cookies.split(';') {
-        let cookie = cookie.trim_start();
-        if cookie.len() < (name.len() + 1 + NONCE_LEN + TAG_LEN)
-            || !cookie.starts_with(name)
-            || cookie.as_bytes()[name.len()] != b'='
-        {
+    for value in headers.get_all("cookie") {
+        let Ok(value) = str::from_utf8(value.as_ref()) else {
```
Comment (djc) [inline mendes/src/cookies.rs:162]: Ah sorry, I forget that this is relatively new. Let's use a `match` for now.

```rust
let value = match str::from_utf8(value.as_ref()) {
    Ok(value) => value,
    Err(_) => continue,
};
```

File: mendes/src/cookies.rs [hunk 6]
```diff
@@ -144,40 +144,44 @@ where
...
             continue;
-        }
+        };
+        // If we have a HTTP/1.1 Cookie header, this single header can contain multiple cookies
```
[the remark this answers is not in the document]
Comment (djc) [reply mendes/src/cookies.rs:165]: That would require a bunch of extra allocations which I'm guessing they might not like to have.

File: mendes/src/cookies.rs [hunk 7]
```diff
@@ -144,9 +144,19 @@ where
...
+            .ok_or(Error::ExpiryWindowTooLong)?;
+
+        let mut bytes = bincode::serialize(&Cookie { expires, data })?;
+        key.encrypt(T::NAME.as_bytes(), &mut bytes)?;
+        Ok(BASE64URL_NOPAD.encode(&bytes))
+    }
+}
+
 fn extract<T: CookieData>(key: &Key, headers: &HeaderMap) -> Option<T> {
-    let cookies = headers.get("cookie")?;
```
Comment (djc) [inline mendes/src/cookies.rs:148]: This doesn't belong in this commit.

File: mendes/src/cookies.rs [hunk 8]
```diff
@@ -158,22 +158,31 @@ impl<T: CookieData> Cookie<T> {
...
-        let cookie = cookie.trim_start();
-        if cookie.len() < (name.len() + 1 + NONCE_LEN + TAG_LEN)
-            || !cookie.starts_with(name)
-            || cookie.as_bytes()[name.len()] != b'='
-        {
-            continue;
-        }
+    // HTTP/2 allows for multiple cookie headers.
+    // https://datatracker.ietf.org/doc/html/rfc9113#name-compressing-the-cookie-head
+    for value in headers.get_all("cookie") {
```
Comment (djc) [inline mendes/src/cookies.rs:163]: Would you mind also importing `http::header::COOKIE` for this, instead?

File: mendes/src/cookies.rs [hunk 9]
```diff
@@ -231,3 +239,62 @@ pub enum Error {
...
+#[cfg(test)]
+mod test {
+    use http::{header, HeaderMap};
+    use serde::{Deserialize, Serialize};
+
+    use super::*;
+
+    /// This test checks that we can extract a cookie from a request that uses multiple cookies in a single header
+    #[test]
+    fn test_http1_cookies() {
```
Comment (djc) [inline mendes/src/cookies.rs:252]: Change the test names to be about the mechanism, not the version?

**Review by djc — COMMENTED** [review]
Sorry for all the nitpicking.

File: mendes/src/cookies.rs (hunk 7 above)
Comment (valkum) [reply mendes/src/cookies.rs:148]: Hmm weird interface glitch.

File: mendes/src/cookies.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (valkum) [reply mendes/src/cookies.rs]: Going to start a discussion over at h2, but let's merge this in the meantime.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Published 0.3.2.
